src/msgtree.rs

```diff
@@ -0,0 +1,178 @@
+//   git-dit - the distributed issue tracker for git
+//   Copyright (C) 2016 Matthias Beyer <mail@beyermatthias.de>
+//   Copyright (C) 2016 Julian Ganz <neither@nut.email>
+//
+//   This program is free software; you can redistribute it and/or modify
+//   it under the terms of the GNU General Public License version 2 as
+//   published by the Free Software Foundation.
+//
+
+use git2::{Commit, Oid};
+use std::fmt::{self, Write};
+use std::iter::FromIterator;
+
+
+/// Representation of graph elements used to display trees
+///
+#[derive(Clone, PartialEq)]
+pub enum TreeGraphElem {
+    Empty,
+    Following,
+    Mark(MarkType),
+}
+
+/// Representation of the type of mark
+///
+#[derive(Clone, PartialEq)]
+pub enum MarkType {
+    Start,
+    Mid,
+    End
+}
+
+
+/// Representation of one line of tree graph elements
+///
+pub struct TreeGraphElemLine(Vec<TreeGraphElem>);
+
+impl TreeGraphElemLine {
+    /// Append a graph element to the line
+    pub fn append(&mut self, e: TreeGraphElem) {
+        self.0.push(e);
+    }
+
+    /// Reverse "start" and "end" marks
+    ///
+    pub fn reverse_marks(&mut self) {
+        for elem in self.0.iter_mut() {
+            match elem {
+                &mut TreeGraphElem::Mark(MarkType::Start) => *elem = TreeGraphElem::Mark(MarkType::End),
+                &mut TreeGraphElem::Mark(MarkType::End) => *elem = TreeGraphElem::Mark(MarkType::Start),
+                _ => {},
+            }
+        }
+    }
+
+    pub fn commit_iterator(self) -> CommitTreeGraphLines {
+        CommitTreeGraphLines(self.0)
+    }
+}
+
+impl FromIterator<TreeGraphElem> for TreeGraphElemLine {
+    fn from_iter<I>(iter: I) -> Self
+        where I: IntoIterator<Item=TreeGraphElem>
+    { TreeGraphElemLine(iter.into_iter().collect()) }
+}
+
+impl fmt::Display for TreeGraphElemLine {
+    fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
+        let mut has_mark = false;
+        for elem in self.0.iter() {
+            f.write_char(match (has_mark, elem) {
+                (_,     &TreeGraphElem::Mark(_)  ) => {has_mark = true; '+'},
+                (true,  _                        ) => '-',
+                (false, &TreeGraphElem::Empty    ) => ' ',
+                (false, &TreeGraphElem::Following) => '|',
+            })?;
```

Comment:
Suggestion:

* Add `is-match` as dependency
* Implementing `to_str()` or something similar on `TreeGraphElem`
* Implementing this whole match like this:

```
{
    if is_match!(elem, &TreeGraphElem::Mark(_)) { has_mark = true; }
    if has_mark { '-' } else { elem.to_s() }
}
```